content_scripts/mode.coffee

```diff
@@ -0,0 +1,210 @@
+# Modes.
+#
+# A mode implements a number of keyboard event handlers which are pushed onto the handler stack when the mode
+# starts, and poped when the mode exits.  The Mode base class takes as single argument options which can
+# define:
+#
+# name:
+#   A name for this mode.
+#
+# badge:
+#   A badge (to appear on the browser popup) for this mode.
+#   Optional.  Define a badge is the badge is constant.  Otherwise, do not define a badge and override the
+#   chooseBadge method instead.  Or, if the mode *never* shows a badge, then do neither.
```

Comment:
There's a typo which makes this hard to parse. Providing example of which modes use which combo of behaviors will make this design clear.